src/prompt/variable.rs

```diff
@@ -146,27 +402,111 @@ pub struct Json;
 
 impl VariableInput for Json {
     fn type_name(&self) -> &str { "json" }
-    fn parse(&self, input: &str) -> Result<Value, Error> {
-        match serde_json::from_str::<serde_json::Value>(input) {
-            Err(e) if e.classify()  == serde_json::error::Category::Eof
-            => Err(Error::Incomplete),
-            Err(e) => Err(no_pos_err(e)),
-            Ok(_) => Ok(Value::Json(
-                model::Json::new_unchecked(input.into())
-            )),
-        }
+    fn parse<'a>(&self, input: &'a str, flags: InputFlags) -> ParseResult<'a> {
+        context(
+            "json",
+            |s: &'a str| {
+                let de = serde_json::Deserializer::from_str(s);
+                let mut stream = de.into_iter::<serde_json::Value>();
+
+                // let the stream do the tokenization gradually until we hit EOF or a syntax error
```

Comment:
I think that the right thing to do here is to *always* do exactly one iteration of the loop.

If there is anything not consumed in the input, that will be an error up at a higher level, right, so there is no need to handle that here?

I think this should try to always parse exactly one json value, and always validate that it parsed correctly, and then leave any remainder to be handled by calling code.

(Either to be parsed, if it is an array or a tuple, or to be errored out on, if this was a top level value)